FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+                    }
+                    catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.UpdateEntityPreconditionFailedErrorCode))
+                    {
+                        _logger.LogWarning(ex, "Update job info entity conflicts.");
+                    }
+                }
+
+                jobInfos.Add(jobInfoEntity.ToJobInfo<TJobInfo>());
+            }
+
+            _logger.LogInformation($"Enqueue jobs '{string.Join(",", jobInfos.Select(jobInfo => jobInfo.Id).ToList())}' successfully.");
+
+            return jobInfos;
+        }
+
+        public async Task<JobInfo> DequeueAsync(byte queueType, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Start to dequeue.");
+
+            // step 1: receive message from message queue
+            TimeSpan? visibilityTimeOut = heartbeatTimeoutSec <= 0 ? null : TimeSpan.FromSeconds(heartbeatTimeoutSec);
+            var message = await _azureJobMessageQueueClient.ReceiveMessageAsync(visibilityTimeOut, cancellationToken);
+
+            if (message.Value == null)
+            {
+                _logger.LogWarning("Failed to receive message from queue, the queue message is null.");
+                return null;
+            }
+
+            var jobMessage = JsonConvert.DeserializeObject<JobMessage>(message.Value.Body.ToString());
+
+            if (jobMessage == null)
+            {
+                _logger.LogWarning("Failed to deserialize message, the deserialized job message is null.");
+                return null;
+            }
+
+            // step 2: get jobInfo entity to check job status, delete this message if job is already completed/failed/cancelled
+            // if status is running and CancelRequest is true, the job will be dequeued, and jobHosting will continue to handle it
+            var jobInfoEntityResponse = await _azureJobInfoTableClient.GetEntityAsync<JobInfoEntity>(
+                jobMessage.PartitionKey,
+                jobMessage.RowKey,
+                cancellationToken: cancellationToken);
+            var jobInfo = jobInfoEntityResponse.Value.ToJobInfo<TJobInfo>();
+
+            if (jobInfo.Status is JobStatus.Completed or JobStatus.Failed or JobStatus.Cancelled)
+            {
+                _logger.LogWarning($"Discard queue message {message.Value.MessageId}, the job status is {jobInfo.Status}.");
+                await _azureJobMessageQueueClient.DeleteMessageAsync(message.Value.MessageId, message.Value.PopReceipt, cancellationToken);
+
+                // dequeue next job
+                return await DequeueAsync(queueType, worker, heartbeatTimeoutSec, cancellationToken);
+            }
+
+            // step 3: get job lock entity to check message id
+            var jobLockEntityResponse = await _azureJobInfoTableClient.GetEntityAsync<JobLockEntity>(
+                jobMessage.PartitionKey,
```

Comment:
We can merge this query with previous one since they are same partition

Reply:
Thanks, I will refine it.

Reply:
To merge get operation, we need to use generic TableEntity instead of JobInfoEntity and JobLockEntity, a https://github.com/microsoft/FHIR-Analytics-Pipelines/pull/153 is created for it.